### ISSUE TITLE
Add check(or AutoFix) TxP core preferences

### ISSUE THREAD

**Issue by makss** [textpattern/textpattern#777]
I propose to add a check basic core preferences (`txp_prefs`) to `Diagnostic` tab.
Can make and AutoFix missing preferences.

All you need we already have:
1. Move array `$prefs = array_merge_recursive(array(...));` from `/textpattern/setup/txpsql.php` to new file `/textpattern/lib/prefs.php`
2. Include `/textpattern/lib/prefs.php` in `/textpattern/setup/txpsql.php`
3. Include `/textpattern/lib/prefs.php` in `/textpattern/include/txp_diag.php`
4. Create new function AutoFix in `txp_diag.php`:  Create prefs if not exist

**Comment by makss** [comment]
Can do the checking and recovery after login to the admin interface with admin privileges.
This will reduce the amount of code in `/textpattern/update/` folder and in setup stage too.

Bonus: We work with the core prefs will be in one place vs many files(update/setup)/functions(autocreate in function)

**Comment by bloatware** [comment]
Let's postpone the discussion to 4.7.

**Comment by bloatware** [comment]
@makss would you mind submitting a patch?

**Comment by Bloke** [comment]
@makss And if you do submit a patch, would you please think about the following:

As part of the Themes implementation, some themes might require certain values for existing prefs or to install new prefs. So, during theme import/export if there was a core API method to set prefs from a file bundled with the theme, that'd be brilliant. This seems to have some crossover with this issue.

Bundling of prefs with themes was not slated to hit 4.7.0, but if we have a mechanism to import them available in core, it might make it viable to roll in support alongside the basic themes implementation. Thanks!

**Comment by makss** [comment]
Another idea was: it would be good to make the structure of sql tables in a separate file, for example `/textpattern/lib/tables.php`

```
$txp_tables = array();

$txp_tables['txp_category'] = "
    id          INT          NOT NULL AUTO_INCREMENT,
    name        VARCHAR(64)  NOT NULL DEFAULT '',
    type        VARCHAR(64)  NOT NULL DEFAULT '',
    parent      VARCHAR(64)  NOT NULL DEFAULT '',
    lft         INT          NOT NULL DEFAULT '0',
    rgt         INT          NOT NULL DEFAULT '0',
    title       VARCHAR(255) NOT NULL DEFAULT '',
    description VARCHAR(255) NOT NULL DEFAULT '',
    PRIMARY KEY (id)
";

$txp_tables['txp_css'] = "
    name VARCHAR(255) NOT NULL,
    css  MEDIUMTEXT   NOT NULL,
    UNIQUE name (name(250))
";
...
```

Then it will be possible to do
```
foreach ($txp_tables as $table=>$txt) {
    safe_create($table, $txt);
}
```
or safe_optimize, check, reCreate some table if need.

**Comment by Bloke** [comment]
@makss Great work so far, thanks. I like the idea of putting the the table definitions in a separate file. We've separated out the template stuff already - though these might be moving to the `themes` folder in due course - so it makes sense to do the same with the table structure.

**Comment by makss** [comment]
@Bloke, I also thought about a separate folder. Where is the best location to place and how to call it?
`/textpattern/update/sql/`
`/textpattern/update/tables/`
`/textpattern/update/structure/`
`/textpattern/lib/sql/`

**Comment by Bloke** [comment]
I prefer `tables` or  `structure`. I feel that `sql` is tied too closely to a specific language.

As to where it sits in our directory structure, I don't mind as long as it makes sense. `setup` is ideal since it is only supposed to be used for initial setup. But if we intend to offer 'repair' or 'optimize' or 'reinstall' or something then it's probably better in `update`. The `lib` folder will probably be (mostly) phased out as we move more towards class-based code in our `vendors` directory. At least, for everything except stuff we need before the autoloader has kicked in.

**Comment by makss** [comment]
I suggest a file name:
`/textpattern/update/structure/txp_category.table`
```
    id          INT          NOT NULL AUTO_INCREMENT,
    name        VARCHAR(64)  NOT NULL DEFAULT '',
    type        VARCHAR(64)  NOT NULL DEFAULT '',
    parent      VARCHAR(64)  NOT NULL DEFAULT '',
    lft         INT          NOT NULL DEFAULT '0',
    rgt         INT          NOT NULL DEFAULT '0',
    title       VARCHAR(255) NOT NULL DEFAULT '',
    description VARCHAR(255) NOT NULL DEFAULT '',

    PRIMARY KEY (id)
```
(BTW, in this table are missing indexes, but it will correct later)

.

Perhaps in the future to make this folder the initial mandatory data. Now it is four rows in `txp_categoriy` and one row in `txp_section`.

`/textpattern/update/structure/txp_category.data`
The contents of the file something like:
```
(1, 'root', 'article', '', 1, 8, 'root', '')
(2, 'root', 'link', '', 1, 4, 'root', '')
(3, 'root', 'image', '', 1, 4, 'root', '')
(4, 'root', 'file', '', 1, 2, 'root', '')
```

`/textpattern/update/structure/txp_section.data`
```
('default', 'default', 'default', '', 1, 1, 1, 'Default')
```

**Comment by Bloke** [comment]
Sounds good to me!

**Comment by makss** [comment]
Can be closed, implemented in #857

**Comment by philwareham** [comment]
Hurrah! Closing, thanks! 🎉
